csrc/optimization/remove_empty.cpp

```diff
@@ -0,0 +1,477 @@
+
+ protected:
+  using DeadCodeRemover::handle;
+
+  //! If tv is a fusion output, we check whether it is empty and if so, replace
+  //! it with full(). For non-outputs that are not inputs, we simply check that
+  //! the tensor is not provably empty.
+  void handle(TensorView* tv) final {
+    if (tv->isFusionInput()) {
+      // Skip inputs since they do not have a definition to redefine
+      return;
+    }
+
+    if (tv->isFusionOutput()) {
+      const auto rfactor =
+          TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+      const auto empty_axes = emptyAxes(rfactor);
+      if (!empty_axes.empty()) {
+        std::vector<Val*> shape(rfactor.size());
+        std::transform(
+            rfactor.begin(), rfactor.end(), shape.begin(), [](IterDomain* id) {
+              return id->extent();
+            });
+        for (auto ax : empty_axes) {
+          shape[ax] = fusion()->zeroVal();
+        }
+        auto dtype = tv->getDataType().value();
+        auto new_tv = full(shape, fusion()->zeroVal(dtype), dtype);
+        replaceTV(tv, new_tv);
+      }
+    } else if (allUsesDead(tv)) {
+      // TensorViews that are not Fusion inputs or outputs and which have no
+      // uses are dead, so remove them and skip processing them and their
+      // definition.
+      markDeadAndMaybeRemove(tv);
+    } else if (isTVEmpty(tv)) {
+      // Note that if there empty intermediate tensors with uses that do not
+      // lead to outputs, this check might fail.
+      TORCH_WARN_ONCE(
+          "Found unexpected empty intermediate TensorView ",
+          tv->toString(),
+          ". This TensorView has un-removed uses that might not be used in this Fusion.");
+    }
+  }
+
+  //! Gets a vector of extents for noReduction(tv->getMaybeRFactorDomain())
+  static std::vector<Val*> noReductionShape(TensorView* tv) {
+    std::vector<Val*> shape;
+    for (auto id : TensorDomain::noReductions(tv->getMaybeRFactorDomain())) {
+      shape.push_back(id->extent());
+    }
+    return shape;
+  }
+
+  //! A reduction over empty axes is equal to the initial value of the
+  //! reduction, as if the reduction were written as follows:
+  //!
+  //!   auto result = init_value;
+  //!   for (auto element : reduction_elements) {
+  //!     result = reduction_op(result, element);
+  //!   }
+  //!   return result;
+  //!
+  void handle(ReductionOp* rop) final {
+    auto in = rop->in()->as<TensorView>();
+    auto empty_input_axes =
+        emptyAxes(TensorDomain::noReductions(in->getMaybeRFactorDomain()));
+    if (empty_input_axes.empty()) {
+      // Input is not empty, handle like any other op
+      return;
+    }
+    auto out = rop->out()->as<TensorView>();
+    // The input is empty in some axes. Assert that they are all reduced
+    for (auto ax : empty_input_axes) {
+      auto id = out->getRootDomain().at(ax);
+      // Input rfactor domain positions correspond to output root positions
+      TORCH_INTERNAL_ASSERT(
```

Comment:
should we assert here?! if we are simply reducing on non-empty dimensions, the pass should just leave it along I think.

i.e. leave this to empty branch to remove that.

Reply:
The op should be marked dead in that case since the output will be dead. If it's not marked dead it means we missed an op somewhere upstream, i.e. there is an unhandled `Expr` type that converts an empty tensor to non-empty. I wanted to catch those types of cases with this assert. This is the same idea as with the "warning" (now an assert) above.

Reply:
Good call. sorry I didn't quite catch this in the first place. So a chain with empty tensor leading to output would be cut-off by the last empty tensor replaced by a factory op. And since we do reverse iterate, we shouldn't have run head on into these cases any longer.

Reply:
I think everything looks good. One quick ask is, can we get a test for this case.

`empty_tensor (input) -> a reduction on non-empty axis -> empty_tensor -> a reduction on empty axis -> tensor (output)`